lightning/src/offers/async_receive_offer_cache.rs

```diff
@@ -0,0 +1,411 @@
+	}
+
+	/// Returns a bool indicating whether new offers are needed in the cache.
+	fn needs_new_offers(&self, duration_since_epoch: Duration) -> bool {
+		// If we have fewer than NUM_CACHED_OFFERS_TARGET offers that aren't expiring soon, indicate
+		// that new offers should be interactively built.
+		let num_unexpiring_offers = self
+			.offers
+			.iter()
+			.filter(|offer| {
+				let offer_absolute_expiry = offer.offer.absolute_expiry().unwrap_or(Duration::MAX);
+				let offer_created_at = offer.offer_created_at;
+				let offer_lifespan =
+					offer_absolute_expiry.saturating_sub(offer_created_at).as_secs();
+				let elapsed = duration_since_epoch.saturating_sub(offer_created_at).as_secs();
+
+				// If an offer is in the last 10% of its lifespan, it's expiring soon.
+				elapsed.saturating_mul(10) < offer_lifespan.saturating_mul(9)
+			})
+			.count();
+
+		num_unexpiring_offers < NUM_CACHED_OFFERS_TARGET
+	}
+
+	// Indicates that onion messages requesting new offer paths have been sent to the static invoice
+	// server. Calling this method allows the cache to self-limit how many requests are sent, in case
+	// the server goes unresponsive.
+	pub(super) fn new_offers_requested(&mut self, duration_since_epoch: Duration) {
+		self.offer_paths_request_attempts += 1;
+		self.last_offer_paths_request_timestamp = duration_since_epoch;
+	}
+
+	/// If we haven't sent an offer paths request in a long time, reset the limit to allow more
+	/// requests to be sent out if/when needed.
+	fn check_reset_offer_paths_request_attempts(&mut self, duration_since_epoch: Duration) {
+		let should_reset =
+			self.last_offer_paths_request_timestamp.saturating_add(PATHS_REQUESTS_BUFFER)
+				< duration_since_epoch;
+		if should_reset {
+			self.reset_offer_paths_request_attempts();
+		}
+	}
+
+	fn reset_offer_paths_request_attempts(&mut self) {
+		self.offer_paths_request_attempts = 0;
+		self.last_offer_paths_request_timestamp = Duration::from_secs(0);
+	}
+
+	/// Returns an iterator over the list of cached offers where the invoice is expiring soon and we
+	/// need to send an updated one to the static invoice server.
+	pub(super) fn offers_needing_invoice_refresh(
+		&self, duration_since_epoch: Duration,
+	) -> impl Iterator<Item = (&Offer, Nonce, Duration, &Responder)> {
+		self.offers.iter().filter_map(move |offer| {
+			const ONE_DAY: Duration = Duration::from_secs(24 * 60 * 60);
+
+			if offer.offer.is_expired_no_std(duration_since_epoch) {
+				return None;
+			}
+			if offer.invoice_update_attempts >= MAX_UPDATE_ATTEMPTS {
+				return None;
+			}
+
+			let time_until_invoice_expiry =
+				offer.static_invoice_absolute_expiry.saturating_sub(duration_since_epoch);
+			let time_until_offer_expiry = offer
+				.offer
+				.absolute_expiry()
+				.unwrap_or_else(|| Duration::from_secs(u64::MAX))
+				.saturating_sub(duration_since_epoch);
+
+			// Update the invoice if it expires in less than a day, as long as the offer has a longer
+			// expiry than that.
```

Comment:
Question for reviewers: I'm not sure it makes sense to only update a static invoice if it expires in less than a day. That doesn't seem like enough time. But it only lives for 2 weeks anyway... Should we maybe make it longer lived, or replace it once it expires in a week, or?